### PR TITLE
Add gtest for Limit, TopN, Projection (#5187)

### LINKED ISSUE #5187: Add gtest for Limit, TopN and Projection executors
## Enhancement
Add gtest for Limit, TopN and Projection executors

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #5187 

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/tests/gtest_executor.cpp [hunk 1]
```diff
@@ -85,6 +85,150 @@ try
...
+    context.addMockTable({db_name, table_name},
+                         {{col_names[0], TiDB::TP::TypeString},
+                          {col_names[1], TiDB::TP::TypeString},
+                          {col_names[2], TiDB::TP::TypeString}},
+                         {toNullableVec<String>(col_names[0], col0),
+                          toNullableVec<String>(col_names[1], col1),
+                          toNullableVec<String>(col_names[2], col2)});
+
+    const size_t concurrency = 1;
+    auto get_request = [&](MockColumnNames col_names)
```
Comment (ywqzzy) [inline dbms/src/Flash/tests/gtest_executor.cpp:207]: Maybe you can make the function in testUtils.

File: dbms/src/Flash/tests/gtest_executor.cpp [hunk 2]
```diff
@@ -85,6 +85,150 @@ try
...
+                         {toNullableVec<String>(col_name, col0)});
+    
+    size_t concurrency = 1;
+    std::shared_ptr<tipb::DAGRequest> request;
+    ColumnsWithTypeAndName expect_cols;
+    
+    bool is_desc;
+
+    /// Check topn result with various parameters
+    for (size_t i = 1; i <= 1; i++)
```
Comment (ywqzzy) [inline dbms/src/Flash/tests/gtest_executor.cpp:157]:
```suggestion
    for (size_t i = 1; i <= 1; ++i)
```

File: dbms/src/Flash/tests/gtest_executor.cpp [hunk 3]
```diff
@@ -85,6 +85,150 @@ try
...
+    const String db_name("test_db");
+    const String table_name("projection_test_table");
+    const String col_name("limit_col");
+
+    /// Create table for the test
+    context.addMockTable({db_name, table_name},
+                         {{col_name, TiDB::TP::TypeString}},
+                         {toNullableVec<String>(col_name, col0)});
+    
+    size_t concurrency = 1;
```
Comment (ywqzzy) [inline dbms/src/Flash/tests/gtest_executor.cpp:107]: the default concurrency is 1

File: dbms/src/Flash/tests/gtest_executor.cpp [hunk 4]
```diff
@@ -85,6 +85,150 @@ try
...
+    std::shared_ptr<tipb::DAGRequest> request;
+    ColumnsWithTypeAndName expect_cols;
+
+    /// Check limit result with various parameters
+    for (size_t limit_num = 0; limit_num <= col_data_num + 3; limit_num++)
```
Comment (ywqzzy) [inline dbms/src/Flash/tests/gtest_executor.cpp:112]:
```suggestion
    for (size_t limit_num = 0; limit_num <= col_data_num + 3; ++limit_num)
```

**Review by ywqzzy — COMMENTED** [review]
Maybe it's better to add new files like gtest_limit_executor.cpp, gtest_projection_executor.cpp and gtest_topn_executor.cpp.

File: dbms/src/Flash/tests/gtest_executor.cpp [hunk 5]
```diff
@@ -85,6 +85,150 @@ try
...
+    {
+        request = context
```
Comment (ywqzzy) [inline dbms/src/Flash/tests/gtest_executor.cpp:114]: Maybe wrap into a function like get_request() below?

File: dbms/src/Flash/tests/gtest_executor.cpp [hunk 6]
```diff
@@ -85,6 +85,150 @@ try
...
+    {
+        is_desc = static_cast<bool>(i); /// Set descent or ascent
+        if (is_desc)
+            sort(col0.begin(), col0.end(), std::greater<ColDataType>()); /// Sort col0 for the following comparison
+        else
+            sort(col0.begin(), col0.end());
+
+        for (size_t limit_num = 0; limit_num <= col_data_num + 3; limit_num++)        
+        {
+            request = context
```
Comment (ywqzzy) [inline dbms/src/Flash/tests/gtest_executor.cpp:167]: wrap it into a function?

File: dbms/src/Flash/tests/gtest_executor.cpp [hunk 7]
```diff
@@ -85,6 +85,150 @@ try
...
+    /// Check topn result with various parameters
+    for (size_t i = 1; i <= 1; i++)
+    {
+        is_desc = static_cast<bool>(i); /// Set descent or ascent
+        if (is_desc)
+            sort(col0.begin(), col0.end(), std::greater<ColDataType>()); /// Sort col0 for the following comparison
+        else
+            sort(col0.begin(), col0.end());
+
+        for (size_t limit_num = 0; limit_num <= col_data_num + 3; limit_num++)
```
Comment (ywqzzy) [inline dbms/src/Flash/tests/gtest_executor.cpp:165]: ditto

File: dbms/src/TestUtils/FunctionTestUtils.cpp [hunk 8]
```diff
@@ -108,8 +108,10 @@ void blockEqual(
     const Block & actual)
 {
     size_t columns = actual.columns();
+    size_t expected_columns = expected.columns();
 
-    ASSERT_TRUE(expected.columns() == columns);
+    ASSERT_TRUE(expected_columns == columns)
+        << std::string("expected: " + std::to_string(expected_columns) + ", actual:" + std::to_string(columns));
```
Comment (SeaRise) [inline dbms/src/TestUtils/FunctionTestUtils.cpp:114]: use `fmt::format` instead?

Comment (xzhangxian1008) [reply dbms/src/TestUtils/FunctionTestUtils.cpp]: Done

File: dbms/src/Flash/tests/gtest_limit_executor.cpp [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        request = getRequest(limit_num);
+
+        if (limit_num == 0)
+            expect_cols = {};
+        else if (limit_num > col_data_num)
+            expect_cols = {toNullableVec<String>(col_name, ColumnWithData(col0.begin(), col0.end()))};
+        else
+            expect_cols = {toNullableVec<String>(col_name, ColumnWithData(col0.begin(), col0.begin() + limit_num))};
+
+        executeStreams(request, expect_cols);
```
[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/tests/gtest_limit_executor.cpp:69]: and add some corner cases?
- the type of limit_num is int, and we can test the case where limit < 0. https://github.com/pingcap/tiflash/blob/69cbfdf8a6bfb1d98ac76dea6e70d87ab3a1ed84/dbms/src/TestUtils/mockExecutor.h#L76
- limit_num = Int.MAX_VALUE.

File: dbms/src/Flash/tests/gtest_projection_executor.cpp [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+};
+
+TEST_F(ExecutorProjectionTestRunner, Projection)
+try
+{
+    auto request = getRequest({col_names[2]});
+    executeStreams(request, {toNullableVec<String>(col_names[2], col2)});
+
+    request = getRequest({col_names[0], col_names[1]});
+    executeStreams(request,
```
Comment (SeaRise) [inline dbms/src/Flash/tests/gtest_projection_executor.cpp:64]: Do we need to test different degrees of parallelism?

File: dbms/src/Flash/tests/gtest_projection_executor.cpp [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    /// Prepare some names
+    std::vector<String> col_names{"proj_col0", "proj_col1", "proj_col2"};
+    const String db_name{"test_db"};
+    const String table_name{"projection_test_table"};
+};
+
+TEST_F(ExecutorProjectionTestRunner, Projection)
+try
+{
+    auto request = getRequest({col_names[2]});
```
Comment (SeaRise) [inline dbms/src/Flash/tests/gtest_projection_executor.cpp:60]: add and test more function in projection?
https://github.com/pingcap/tiflash/blob/69cbfdf8a6bfb1d98ac76dea6e70d87ab3a1ed84/dbms/src/TestUtils/mockExecutor.h#L176-L189
https://github.com/pingcap/tiflash/blob/69cbfdf8a6bfb1d98ac76dea6e70d87ab3a1ed84/dbms/src/Debug/astToExecutor.cpp#L123
https://github.com/pingcap/tiflash/blob/69cbfdf8a6bfb1d98ac76dea6e70d87ab3a1ed84/dbms/src/Debug/astToExecutor.cpp#L309

Comment (SeaRise) [reply dbms/src/Flash/tests/gtest_projection_executor.cpp:60]: and add some corner cases?
such as
- project({})
- projec("a1", "a1", "a1"), same cols in projection
- col num in projection is 100 or more?

File: dbms/src/Flash/tests/gtest_topn_executor.cpp [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+        for (size_t limit_num = 0; limit_num <= col_data_num + 3; limit_num++)
+        {
+            request = getRequest(col_name, is_desc, limit_num);
+
+            if (limit_num == 0 || limit_num > col_data_num)
+                expect_cols = {toNullableVec<String>(col_name, ColumnWithData(col0.begin(), col0.end()))};
+            else
+                expect_cols = {toNullableVec<String>(col_name, ColumnWithData(col0.begin(), col0.begin() + limit_num))};
+
+            executeStreams(request, expect_cols);
```
Comment (SeaRise) [inline dbms/src/Flash/tests/gtest_topn_executor.cpp:76]: Do we need to test different degrees of parallelism?

File: dbms/src/Flash/tests/gtest_topn_executor.cpp [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    /// Check topn result with various parameters
+    for (size_t i = 1; i <= 1; ++i)
+    {
+        is_desc = static_cast<bool>(i); /// Set descent or ascent
+        if (is_desc)
+            sort(col0.begin(), col0.end(), std::greater<ColDataType>()); /// Sort col0 for the following comparison
+        else
+            sort(col0.begin(), col0.end());
+
+        for (size_t limit_num = 0; limit_num <= col_data_num + 3; limit_num++)
```
Comment (SeaRise) [inline dbms/src/Flash/tests/gtest_topn_executor.cpp:67]: and add some corner cases like limit?

File: dbms/src/Flash/tests/gtest_topn_executor.cpp [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    void initializeContext() override
+    {
+        ExecutorTest::initializeContext();
+
+        context.addMockTable({db_name, table_name},
+                             {{col_name, TiDB::TP::TypeString}},
+                             {toNullableVec<String>(col_name, col0)});
+    }
+
+    std::shared_ptr<tipb::DAGRequest> getRequest(const String & col_name, bool is_desc, int limit_num)
```
Comment (SeaRise) [inline dbms/src/Flash/tests/gtest_topn_executor.cpp:38]: add more corner cases?
such as 
- 100 or more order by items
- function in order by items

File: dbms/src/Flash/tests/gtest_topn_executor.cpp (hunk 13 above)
Comment (SeaRise) [reply dbms/src/Flash/tests/gtest_topn_executor.cpp:67]: use `++limit_num` instead

File: dbms/src/Flash/tests/gtest_projection_executor.cpp (hunk 11 above)
Comment (ywqzzy) [reply dbms/src/Flash/tests/gtest_projection_executor.cpp:60]: > and add complex function? such as `concat(concat(concat(..............................)))`

Can ref to https://github.com/pingcap/tiflash/pull/5143 to add concat function

File: dbms/src/Flash/tests/gtest_projection_executor.cpp (hunk 10 above)
Comment (ywqzzy) [reply dbms/src/Flash/tests/gtest_projection_executor.cpp:64]: > Do we need to test different degrees of parallelism?

If test with parellelism, some executors need to pay extra attention to the stability of test result. For now, the three executors will have stable test result. I think it will be ok to test different degress of parallelism.

File: dbms/src/Flash/tests/gtest_limit_executor.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Flash/tests/gtest_limit_executor.cpp:69]: > 

limit < 0 is invalid in TiDB and will be raised as SQL syntax error when user try to input this parameter. So I think it's needless to test this case.

[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Flash/tests/gtest_limit_executor.cpp:69]: > and add some corner cases?
> 
> * the type of limit_num is int, and we can test the case when limit < 0. https://github.com/pingcap/tiflash/blob/69cbfdf8a6bfb1d98ac76dea6e70d87ab3a1ed84/dbms/src/TestUtils/mockExecutor.h#L76
> * limit_num = Int.MAX_VALUE.

limit < 0 is invalid in TiDB and will be raised as SQL syntax error when user try to input this parameter. So I think it's needless to test this case.

File: dbms/src/TestUtils/FunctionTestUtils.cpp [hunk 15]
```diff
@@ -108,8 +108,10 @@ void blockEqual(
     const Block & actual)
 {
     size_t columns = actual.columns();
+    size_t expected_columns = expected.columns();
 
-    ASSERT_TRUE(expected.columns() == columns);
+    ASSERT_TRUE(expected_columns == columns)
+        << fmt::format("Expected: {}, Actual: {}", expected_columns, columns);
```
Comment (JaySon-Huang) [inline dbms/src/TestUtils/FunctionTestUtils.cpp:114]:
```suggestion
    ASSERT_EQ(expected_columns, columns);
```
Use `ASSERT_EQ` and it will show the actual number when they don't match. And we don't need to manually write the message to show the value.

Comment (Willendless) [reply dbms/src/TestUtils/FunctionTestUtils.cpp:114]: Maybe we can use something like

```cpp
::testing::AssertionResult blockEqual(
    const Block & expected,
    const Block & actual)
{
    size_t columns = actual.columns();

    ASSERT_EQUAL(expected.columns(), columns, "Block size mismatch");

    for (size_t i = 0; i < columns; ++i)
    {
        const auto & expected_col = expected.getByPosition(i);
        const auto & actual_col = actual.getByPosition(i);
        auto ret = columnEqual(expected_col, actual_col);
        if (!ret)
            return ret;
    }

    return ::testing::AssertionSuccess();
}
```

File: dbms/src/Flash/tests/gtest_projection_executor.cpp (hunk 10 above)
Comment (SeaRise) [reply dbms/src/Flash/tests/gtest_projection_executor.cpp:64]: we can use `TopN` to  ensure the order of result rows like https://github.com/pingcap/tiflash/blob/master/dbms/src/Flash/tests/gtest_executor.cpp#L91-L95

File: dbms/src/TestUtils/FunctionTestUtils.cpp (hunk 15 above)
Comment (SeaRise) [reply dbms/src/TestUtils/FunctionTestUtils.cpp:114]: > Maybe we can use something like
> 
> ```c++
> ::testing::AssertionResult blockEqual(
>     const Block & expected,
>     const Block & actual)
> {
>     size_t columns = actual.columns();
> 
>     ASSERT_EQUAL(expected.columns(), columns, "Block size mismatch");
> 
>     for (size_t i = 0; i < columns; ++i)
>     {
>         const auto & expected_col = expected.getByPosition(i);
>         const auto & actual_col = actual.getByPosition(i);
>         auto ret = columnEqual(expected_col, actual_col);
>         if (!ret)
>             return ret;
>     }
> 
>     return ::testing::AssertionSuccess();
> }
> ```

it seems to same as https://github.com/pingcap/tiflash/blob/73e708cd22b935ca240a236a87e261aabddd770e/dbms/src/TestUtils/FunctionTestUtils.cpp#L106-L121

Comment (Willendless) [reply dbms/src/TestUtils/FunctionTestUtils.cpp:114]: Oh yeah, it is just a little refactoring. The advantage might lies in we have better output error message when column size is not expected.

File: dbms/src/Flash/tests/gtest_projection_executor.cpp (hunk 10 above)
Comment (xzhangxian1008) [reply dbms/src/Flash/tests/gtest_projection_executor.cpp:74]: > we can use `TopN` to ensure the order of result rows like
> 
> https://github.com/pingcap/tiflash/blob/73e708cd22b935ca240a236a87e261aabddd770e/dbms/src/Flash/tests/gtest_executor.cpp#L91-L95

get, I will try it

Comment (xzhangxian1008) [reply dbms/src/Flash/tests/gtest_projection_executor.cpp:187]: > we can use `TopN` to ensure the order of result rows like
> 
> https://github.com/pingcap/tiflash/blob/73e708cd22b935ca240a236a87e261aabddd770e/dbms/src/Flash/tests/gtest_executor.cpp#L91-L95

Done

File: dbms/src/Flash/tests/gtest_limit_executor.cpp [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+    void initializeContext() override
+    {
+        ExecutorTest::initializeContext();
+
+        context.addMockTable({db_name, table_name},
+                             {{col_name, TiDB::TP::TypeString}},
+                             {toNullableVec<String>(col_name, col0)});
+    }
+
+    std::shared_ptr<tipb::DAGRequest> getRequest(size_t limit_num)
```
Comment (ywqzzy) [inline dbms/src/Flash/tests/gtest_limit_executor.cpp:38]: move the function into ExecutorTestUtils and rename it into a more explicit name？

File: dbms/src/Flash/tests/gtest_projection_executor.cpp [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+try
+{
+    /// Check single column
+    auto request = getRequest<MockColumnNames>({col_names[4]}, col_names[4]);
+    executeWithConcurrency(request, {toNullableVec<Int32>(col_names[4], col4_sorted_asc)});
+
+    /// Check multi columns
+    request = getRequest<MockColumnNames>({col_names[0], col_names[4]}, col_names[4]);
+    executeWithConcurrency(request,
+                           {
```
Comment (ywqzzy) [inline dbms/src/Flash/tests/gtest_projection_executor.cpp:96]: code style, maybe "{" should in the same line with "toNullableVec"

File: dbms/src/Flash/tests/gtest_topn_executor.cpp [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+                              {col_name[1], TiDB::TP::TypeString},
+                              {col_name[2], TiDB::TP::TypeString},
+                              {col_name[3], TiDB::TP::TypeLong}},
+                             {toNullableVec<Int32>(col_name[0], col_age),
+                              toNullableVec<String>(col_name[1], col_gender),
+                              toNullableVec<String>(col_name[2], col_country),
+                              toNullableVec<Int32>(col_name[3], c0l_salary)});
+    }
+
+    std::shared_ptr<tipb::DAGRequest> getRequest(const String & table_name, const String & col_name, bool is_desc, int limit_num)
```
Comment (ywqzzy) [inline dbms/src/Flash/tests/gtest_topn_executor.cpp:50]: ditto

File: dbms/src/Flash/tests/gtest_topn_executor.cpp [hunk 19]
```diff
@@ -0,0 +1,228 @@
...
+    {
+        return context.scan(db_name, table_name).topN(col_name, is_desc, limit_num).build(context);
+    }
+
+    std::shared_ptr<tipb::DAGRequest> getRequest(const String & table_name, MockOrderByItems order_by_items, int limit, MockAsts func_proj_ast = {}, MockColumnNames out_proj_ast = {})
```
Comment (ywqzzy) [inline dbms/src/Flash/tests/gtest_topn_executor.cpp:55]: ditto

File: dbms/src/Flash/tests/gtest_topn_executor.cpp [hunk 20]
```diff
@@ -0,0 +1,228 @@
...
+        }
+
+        {
+            /// select * from clerk order by gender DESC, country ASC, salary DESC;
+            order_by_items = {MockOrderByItem(col_name[1], true), MockOrderByItem(col_name[2], false), MockOrderByItem(col_name[3], true)};
+            request = getRequest(table_name, order_by_items, 100);
+            executeStreams(request, expect_cols[2]);
+        }
+    }
+}
```
Comment (ywqzzy) [inline dbms/src/Flash/tests/gtest_topn_executor.cpp:152]:
```suggestion
        MockOrderByItems order_by_items;  // vector<vector<OrderByItem>
        for (i:0...3)
        {
            request = getRequest(table_name, order_by_items[i], 100);
            executeStreams(request, expect_cols[i]);
        }
    }
}
```

File: dbms/src/Flash/tests/gtest_topn_executor.cpp [hunk 21]
```diff
@@ -0,0 +1,228 @@
...
+CATCH
+
+TEST_F(ExecutorTopNTestRunner, TopNFunction)
+try
+{
+    std::shared_ptr<tipb::DAGRequest> request;
+    std::vector<ColumnsWithTypeAndName> expect_cols;
```
Comment (ywqzzy) [inline dbms/src/Flash/tests/gtest_topn_executor.cpp:159]: Maybe not a vector?

File: dbms/src/Flash/tests/gtest_limit_executor.cpp (hunk 16 above)
Comment (xzhangxian1008) [reply dbms/src/Flash/tests/gtest_limit_executor.cpp:38]: > move the function into ExecutorTestUtils and rename it into a more explicit name？

1. Building tipb::DAGRequest is needed frequently in gtests for executor. However, the way of building DAGRequest is definitely different between different executors(Even parameters in getRequest are different from each other, and we can get this conclusion after comparing the implementation of getRequest in gtest_limit_executor, gtest_projection_executor and gtest_topn_executor). 

2. I will rename it into a more explicit name.

File: dbms/src/Flash/tests/gtest_topn_executor.cpp (hunk 21 above)
Comment (xzhangxian1008) [reply dbms/src/Flash/tests/gtest_topn_executor.cpp:159]: > Maybe not a vector?

Considering for the future tests, I think we may need vector in the future to contain more expected cols.

File: dbms/src/Flash/tests/gtest_topn_executor.cpp (hunk 20 above)
Comment (xzhangxian1008) [reply dbms/src/Flash/tests/gtest_topn_executor.cpp:152]: > 

Done

**Review by SeaRise — APPROVED** [review]

**Review by ywqzzy — APPROVED** [review]
LGTM
